### PR TITLE
Started work on self-reconnecting reverse tunnels

### PR DESCRIPTION
**Pull request by kontsevoy**
Better logic for reconnecting on lost connections.

### CODE REVIEW

File: lib/client/keyagent.go [hunk 1]
```diff
@@ -109,9 +109,8 @@ func (a *LocalKeyAgent) CheckHostSignature(hostId string, remote net.Addr, key s
 		return trace.Wrap(err)
 	}
 	for i := range keys {
-		log.Infof("checking host %v vs trusted CA %v for %v", sshutils.Fingerprint(cert.SignatureKey), sshutils.Fingerprint(keys[i]), hostId)
+		//log.Infof("checking host %v vs trusted CA %v for %v", sshutils.Fingerprint(cert.SignatureKey), sshutils.Fingerprint(keys[i]), hostId)
```
Comment (klizhentas) [inline lib/client/keyagent.go]: leftover comment

File: lib/reversetunnel/srv.go [hunk 2]
```diff
@@ -173,35 +173,31 @@ func (s *server) Close() error {
...
-
-		ch, req, err := nch.Accept()
-		if err != nil {
-			log.Errorf("failed to accept channel: %v", err)
-			sconn.Close()
-			return
-		}
-		go site.handleHeartbeat(remoteConn, ch, req)
+	if nch.ChannelType() != chanHeartbeat {
+		log.Warningf("reversetunnel received unknown channel request %v from %v",
```
Comment (klizhentas) [inline lib/reversetunnel/srv.go]: I think here you'd better do `nch.Reject` to tell client the error and not just hang up

**Comment by klizhentas** [comment]
lgtm with couple of suggestions/cleanups

**Comment by kontsevoy** [comment]
done

**Pull request merged**